api/src/main/java/io/minio/MinioClient.java

```diff
@@ -932,43 +1030,49 @@ private void updateRegionCache(String bucketName)
            InternalException {
     if (bucketName != null && this.accessKey != null && this.secretKey != null
         && !BucketRegionCache.INSTANCE.exists(bucketName)) {
-      Map<String,String> queryParamMap = new HashMap<>();
-      queryParamMap.put("location", null);
-
-      HttpResponse response = execute(Method.GET, US_EAST_1, bucketName, null,
-                                      null, queryParamMap, null, 0);
-
-      // existing XmlEntity does not work, so fallback to regular parsing.
-      XmlPullParser xpp = xmlPullParserFactory.newPullParser();
-      String location = null;
-
-      xpp.setInput(response.body().charStream());
-      while (xpp.getEventType() != xpp.END_DOCUMENT) {
-        if (xpp.getEventType() ==  xpp.START_TAG && xpp.getName() == "LocationConstraint") {
+      // Compute bucket region but always pick this.region if the latter is set,
+      // otherwise, call the bucket location API.
+      String region;
+      if (!this.region.equals("")) {
+        region = this.region;
+      } else {
+        Map<String,String> queryParamMap = new HashMap<>();
+        queryParamMap.put("location", null);
+
+        HttpResponse response = execute(Method.GET, US_EAST_1, bucketName, null,
+            null, queryParamMap, null, 0);
+
+        // existing XmlEntity does not work, so fallback to regular parsing.
+        XmlPullParser xpp = xmlPullParserFactory.newPullParser();
+        String location = null;
+
+        xpp.setInput(response.body().charStream());
+        while (xpp.getEventType() != xpp.END_DOCUMENT) {
+          if (xpp.getEventType() ==  xpp.START_TAG && xpp.getName() == "LocationConstraint") {
+            xpp.next();
+            location = getText(xpp, location);
+            break;
+          }
           xpp.next();
-          location = getText(xpp, location);
-          break;
         }
-        xpp.next();
-      }
 
-      // close response body.
-      response.body().close();
+        // close response body.
+        response.body().close();
 
-      String region;
-      if (location == null) {
-        region = US_EAST_1;
-      } else {
-        // eu-west-1 can be sometimes 'EU'.
-        if ("EU".equals(location)) {
-          region = "eu-west-1";
+        if (location == null) {
+          region = US_EAST_1;
         } else {
-          region = location;
+          // eu-west-1 can be sometimes 'EU'.
+          if ("EU".equals(location)) {
+            region = "eu-west-1";
+          } else {
+            region = location;
+          }
         }
       }
 
       // Add the new location.
-      BucketRegionCache.INSTANCE.add(bucketName, region);
+      BucketRegionCache.INSTANCE.set(bucketName, region);
```

Comment:
I thought of BucketRegionCache.region() could return passed region for non-empty bucket name.  I am OK with this approach too.